experimental/oauthtokenretriever/token.go

```diff
@@ -0,0 +1,107 @@
+package oauthtokenretriever
+
+import (
+	"context"
+	"fmt"
+	"net/url"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/google/uuid"
+	"golang.org/x/oauth2"
+	"golang.org/x/oauth2/clientcredentials"
+)
+
+type TokenRetriever interface {
+	OnBehalfOfUser(userID string) (string, error)
+	Self() (string, error)
+}
+
+type tokenRetriever struct {
+	signer signer
+	conf   *clientcredentials.Config
+}
+
+// tokenPayload returns a JWT payload for the given user ID, client ID, and host.
+func (t *tokenRetriever) tokenPayload(userID string) map[string]interface{} {
+	iat := time.Now().Unix()
+	exp := iat + 1800
+	u := uuid.New()
+	payload := map[string]interface{}{
+		"iss": t.conf.ClientID,
+		"sub": fmt.Sprintf("user:id:%s", userID),
+		"aud": t.conf.TokenURL,
+		"exp": exp,
+		"iat": iat,
+		"jti": u.String(),
+	}
+	return payload
+}
+
+func (t *tokenRetriever) Self() (string, error) {
+	t.conf.EndpointParams = url.Values{}
+	tok, err := t.conf.TokenSource(context.Background()).Token()
+	if err != nil {
+		return "", err
+	}
+	return tok.AccessToken, nil
+}
+
+func (t *tokenRetriever) OnBehalfOfUser(userID string) (string, error) {
+	signed, err := t.signer.sign(t.tokenPayload(userID))
+	if err != nil {
+		return "", err
+	}
+
+	t.conf.EndpointParams = url.Values{
+		"grant_type": {"urn:ietf:params:oauth:grant-type:jwt-bearer"},
+		"assertion":  {signed},
+	}
+	tok, err := t.conf.TokenSource(context.Background()).Token()
```

Comment:
Nit. Maybe should take context as argument to `OnBehalfOfUser` rather than providing context.Background()?